BV2-6 (feat) : 사용자 로그인 기능 개발

seob7 opened the pull request:
## #️⃣ Issue Number

N/A

## 📝 요약(Summary)

- 로그인 기능 및 토큰 재발급 기능 개발하였습니다.
    - `refreshToken` -> redis 저장 

- 다음 작업 목록
    - `email 중복 확인 API`
    - `nickName 중복 확인 API`

## 🛠️ PR 유형

어떤 변경 사항이 있나요?

- [x] 새로운 기능 추가
- [ ] 버그 수정
- [ ] 코드 리팩토링
- [ ] 주석 추가 및 수정
- [ ] 문서 수정
- [ ] 테스트 추가, 테스트 리팩토링
- [x] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

## 💬 공유사항 to 리뷰어

둘 중 어떤걸로 해야할지 고민입니다.
`httpSession.setAttribute("refreshToken:" + userId, refreshToken);` OR `httpSession.setAttribute(userId, refreshToken);`

<!--- 리뷰어가 중점적으로 봐줬으면 좋겠는 부분이 있으면 적어주세요. -->
<!--- 논의해야할 부분이 있다면 적어주세요.-->
<!--- ex) 메서드 XXX의 이름을 더 잘 짓고 싶은데 혹시 좋은 명칭이 있을까요? -->

## 📑 API Spec

| **기능**          |     로그인           |
|-----------------|----------------|
| **HTTP Method** |     POST       |
| **URL**         |        /open-api/user/login            |

### Request Body
```
{
  "result": {
    "resultCode": 0,
    "resultMessage": "string",
    "resultDescription": "string"
  },
  "body": {
    "email": "string12@example.com",
    "password": "String12@"
  }
}
```

### Response Body
```
{
    "result": {
        "resultCode": 200,
        "resultMessage": "성공",
        "resultDescription": "성공"
    },
    "body": {
        "accessToken": "string",
        "accessTokenExpiredAt": "2025-01-06T01:03:21.188086",
        "refreshToken": "string",
        "refreshTokenExpiredAt": "2025-01-06T12:03:21.214714"
    }
}
```
---

| **기능**          |      토큰 재발급           |
|-----------------|----------------|
| **HTTP Method** |            |
| **URL**         |        /open-api/user/reissue            |

### Request Header
|  **Key**        |      **Value**           |
|-----------------|----------------|
|  Authorization    |   Bearer stringstringstring |

### Response Body
```
{
    "result": {
        "resultCode": 200,
        "resultMessage": "성공",
        "resultDescription": "성공"
    },
    "body": {
        "token": "string",
        "expiredAt": "2025-01-06T10:20:38.573074"
    }
}
```

## 📸스크린샷 (선택)


## ✅ PR Checklist

PR이 다음 요구 사항을 충족하는지 확인하세요.

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x] 변경 사항에 대한 테스트를 했습니다. (버그 수정/기능에 대한 테스트)

seob7 committed:
BV2-23 (feat) : 토큰 관리 기능 개발

seob7 committed:
BV2-6 (feat) : 사용자 로그인 기능 개발

BV2-6 (feat) : 사용자 로그인 기능 개발

seob7 committed:
BV2-24 (feat) : 사용자 로그인 및 토큰 예외처리, 유효성 검사 처리

BV2-24 (feat) : 사용자 로그인 및 토큰 예외처리, 유효성 검사 처리

shinywoon commented on user/src/main/java/user/security/jwt/filter/JwtAuthFilter.java:
```diff
@@ -0,0 +1,59 @@
+
+@Slf4j
+@RequiredArgsConstructor
+public class JwtAuthFilter extends OncePerRequestFilter {
```
AuthFilter 는 API Gateway Module에서 구현 후 등록 및 관리 될 예정입니다.

shinywoon commented on user/src/main/java/user/application/UserRegisterService.java:
```diff
@@ -22,7 +24,8 @@ public boolean register(UserRegisterCommand userRegisterCommand) {
     }
 
-    private static void initializeUser(UserRegisterCommand userRegisterCommand) {
+    private void initializeUser(UserRegisterCommand userRegisterCommand) {
```
`initializeUser`: Method 명이 적절하지 않은 것 같습니다.
기능에 충실한 Method 명을 사용하도록 노력해주세요.
추천: initUserParameter, setMembershipRegistrationParameters

shinywoon commented on user/src/main/java/user/adapter/input/web/UserOpenApiController.java:
```diff
@@ -29,4 +42,18 @@ public Api<Boolean> register(@RequestBody @Valid Api<UserRegisterRequest> userRe
     }
 
+    @PostMapping("/login")
+    public Api<TokenResponse> login(@RequestBody @Valid Api<UserLoginRequest> userLoginRequest) {
+        UserLoginCommand loginCommand = tokenConverter.toLoginCommand(userLoginRequest.getBody());
+        TokenCommand tokenCommand = userLoginUseCase.login(loginCommand);
+        TokenResponse response = tokenConverter.toTokenResponse(tokenCommand);
+        return Api.OK(response);
+    }
```
리뷰의 고민은 아직 구현이 안되어있네요?
고민은 이 포스터 참조바랍니다.
[https://velog.io/@mj3242/Spring-Redis-활용-access-Token-refresh-Token-인증인가-구현](url)

seob7 replied:
해당 코드는 `TokenHelperService` 에 구현돼 있습니다.

<img width="627" alt="스크린샷 2025-01-06 10 11 46" src="https://github.com/user-attachments/assets/8b652579-da06-453a-adb3-387c5f5bfdaa" />

seob7 committed:
BV2-6 (refactor) : PR 요구사항 반영

BV2-6 (refactor) : PR 요구사항 반영

1. JwtAuthFilter 제거, SpringSecurityConfig 수정
2. UserRegisterService 메서드명 수정
    -  initializeUser -> initUserParameter
